cluster/clustered_data.go

```diff
@@ -0,0 +1,199 @@
+			StateUpdatorOpts{}),
+		opts:       opts,
+		readyState: clusteredDataStateReady,
+	}
+}
+
+type ClusteredDataOpts struct {
+	AvailabilityRetryInterval time.Duration
+	ObjStoreCallTimeout       time.Duration
+}
+
+func (mo *ClusteredDataOpts) setDefaults() {
+	if mo.AvailabilityRetryInterval == 0 {
+		mo.AvailabilityRetryInterval = DefaultAvailabilityRetryInterval
+	}
+}
+
+func (m *ClusteredData) LoadData() ([]byte, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.readyState == clusteredDataStateStopped {
+		return nil, errors.New("stopped")
+	}
+	// Atomically increment the epoch
+	buff, err := m.stateMachine.Update(func(state []byte) ([]byte, error) {
+		epoch := buffToEpoch(state)
+		newState := make([]byte, 8)
+		binary.BigEndian.PutUint64(newState, epoch+1)
+		return newState, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	m.epoch = buffToEpoch(buff)
+	prevEpoch := m.epoch - 1
+	// Now we try and load the key with the highest epoch - the key might be lower than prevEpoch as no data might
+	// have been stored in previous epoch
+	var data []byte
+	for {
+		dataKey := m.createDataKey(prevEpoch)
+		data, err = m.getWithRetry(dataKey)
+		if err != nil {
+			return nil, err
+		}
+		if len(data) == 0 {
+			// no key found - try with next lower epoch
+			if prevEpoch == 0 {
+				// No data to load
+				break
+			}
+			prevEpoch--
+		} else {
+			break
+		}
+	}
+	m.readyState = clusteredDataStateLoaded
+	return data, nil
+}
+
+func (m *ClusteredData) StoreData(data []byte) (bool, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.readyState != clusteredDataStateLoaded {
+		return false, errors.New("not loaded")
+	}
+	dataKey := m.createDataKey(m.epoch)
+	if err := m.putWithRetry(dataKey, data); err != nil {
+		return false, err
+	}
+	// Then we get latest epoch using a no-op update
+	buff, err := m.stateMachine.Update(func(buff []byte) ([]byte, error) {
+		return buff, nil
+	})
+	if err != nil {
+		return false, err
+	}
+	currEpoch := buffToEpoch(buff)
+	if currEpoch != m.epoch {
+		// Epoch has changed - fail the store
+		log.Debug("controller failed to store metadata as epoch has changed")
+		return false, nil
+	}
+	return true, nil
+}
+
+func (m *ClusteredData) Stop() {
+	// Allow any retry loops to exit
+	m.stopping.Store(true)
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.readyState = clusteredDataStateStopped
+}
+
+func (m *ClusteredData) getWithRetry(dataKey string) ([]byte, error) {
+	for {
+		if m.stopping.Load() {
+			return nil, errors.New("clustered data is stopping")
+		}
+		data, err := objstore.GetWithTimeout(m.objStoreClient, m.dataBucketName, dataKey, m.opts.ObjStoreCallTimeout)
+		if err == nil {
+			return data, nil
+		}
+		if !common.IsUnavailableError(err) {
+			return nil, err
+		}
+		// retry
+		time.Sleep(m.opts.AvailabilityRetryInterval)
+	}
+}
+
+func (m *ClusteredData) putWithRetry(dataKey string, data []byte) error {
+	for {
+		if m.stopping.Load() {
+			return errors.New("clustered data is stopping")
```

Comment:
`errors.New(fmt.Sprintf("won't put key %s because clustered data is stopping", dataKey))`


?